### PR TITLE
Fixed copy resources from jar when tess4j is used as maven deployed at JBoss/Wildfly

### PR DESCRIPTION
**Pull request by itsura**
Fixed problem with copy resources from jar when tess4j is used as maven dependency in Web App deployed at JBoss/Wildfly.
p.s. Sorry for source formatting my settings are with tabs.

### CODE REVIEW

**Comment by nguyenq** [comment]
It's really difficult to review the actual changes when the entire files were replaced due to differences in identations, line endings... Preferably, we want to see only the code statements that actually changed. Therefore, submitted files should follow the format of the files already in the repostory.

**Comment by 4F2E4A2E** [comment]
@itsura how did you format those files? which ide are you using?

**Comment by itsura** [comment]
I squash commits and do my best to fix formatting. I use eclipse and tried to fix formatting to be like as yours.

**Comment by Gasprai386** [comment]
Will it be resolved?
Or could you provide any alternative solution, to install tess4j as a dependency to war based project deployed in jboss/wildfly?

**Comment by nguyenq** [comment]
The jboss-vfs support has been pulled in https://github.com/nguyenq/tess4j/releases/tag/tess4j-3.4.0 release. Will review the PDF fix when have more time. Thanks.

File: pom.xml [hunk 1]
```diff
@@ -209,6 +209,11 @@
             <artifactId>log4j-over-slf4j</artifactId>
             <version>1.7.21</version>
         </dependency>
+         <dependency>
```
Comment (4F2E4A2E) [inline pom.xml]: @nguyenq the changes of this file are safe and makes sende for me, though crappy format.

File: pom.xml [hunk 2]
```diff
@@ -209,6 +209,11 @@
             <artifactId>log4j-over-slf4j</artifactId>
```
Comment (4F2E4A2E) [inline pom.xml]: @nguyenq the changes of this file are safe and makes sende for me, though crappy format.

File: pom.xml [hunk 3]
```diff
@@ -209,6 +209,11 @@
             <artifactId>log4j-over-slf4j</artifactId>
             <version>1.7.21</version>
         </dependency>
```
Comment (4F2E4A2E) [inline pom.xml]: @nguyenq the changes of this file are safe and makes sende for me, though crappy format.

File: src/main/java/net/sourceforge/tess4j/Tesseract.java [hunk 4]
```diff
@@ -199,22 +212,38 @@ public String doOCR(File imageFile) throws TesseractException {
...
      * Performs OCR operation.
      *
-     * @param imageFile an image file
+     * @param inputFile an image file
      * @param rect the bounding rectangle defines the region of the image to be
      * recognized. A rectangle of zero dimension or <code>null</code> indicates
      * the whole image.
      * @return the recognized text
      * @throws TesseractException
      */
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/Tesseract.java]: @nguyenq the changes of this file are safe and makes sende for me, though he is implementing the logic which every tess4j api consumer should do on his own in order to maintain flexibility, it's your call to decide.

File: src/main/java/net/sourceforge/tess4j/Tesseract1.java [hunk 5]
```diff
@@ -167,22 +180,38 @@ public String doOCR(File imageFile) throws TesseractException {
...
      * Performs OCR operation.
      *
-     * @param imageFile an image file
+     * @param inputFile an image file
      * @param rect the bounding rectangle defines the region of the image to be
      * recognized. A rectangle of zero dimension or <code>null</code> indicates
      * the whole image.
      * @return the recognized text
      * @throws TesseractException
      */
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/Tesseract1.java]: @nguyenq the changes of this file are safe and makes sende for me, though he is implementing the logic which every tess4j api consumer should do on his own in order to maintain flexibility, it's your call to decide.

File: src/main/java/net/sourceforge/tess4j/util/ImageIOHelper.java [hunk 6]
```diff
@@ -58,11 +58,11 @@
 
     private static final Logger logger = LoggerFactory.getLogger(new LoggHelper().toString());
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/util/ImageIOHelper.java]: @nguyenq the changes of this file are safe and makes sende for me, though crappy format.

File: src/main/java/net/sourceforge/tess4j/util/ImageIOHelper.java [hunk 7]
```diff
@@ -105,49 +105,48 @@
 
         ImageReader reader = readers.next();
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/util/ImageIOHelper.java]: @nguyenq the changes of this file are safe and makes sense for me, he did da great error handling job and added nice null checks, though crappy format.

File: src/main/java/net/sourceforge/tess4j/util/ImageIOHelper.java [hunk 8]
```diff
@@ -194,11 +193,11 @@
             }
 
             File tiffFile = File.createTempFile(OUTPUT_FILE_NAME, TIFF_EXT);
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/util/ImageIOHelper.java]: @nguyenq the changes of this file are safe and makes sense for me, though ios is not being closed anymore.

File: src/main/java/net/sourceforge/tess4j/util/ImageIOHelper.java [hunk 9]
```diff
@@ -322,71 +321,23 @@ public static ByteBuffer convertImageData(BufferedImage bi) {
...
-    /**
-     * Gets a list of <code>IIOImage</code> objects for an image file.
-     *
-     * @param imageFile input image file. It can be any of the supported
-     * formats, including TIFF, JPEG, GIF, PNG, BMP, JPEG, and PDF if GPL
-     * Ghostscript is installed
-     * @return a list of <code>IIOImage</code> objects
-     * @throws IOException
-     */
-    public static List<IIOImage> getIIOImageList(File imageFile) throws IOException {
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/util/ImageIOHelper.java]: @nguyenq why is this method being removed? I don't see the need for it, just to write it again? I don't see the need for that in here. It's just another untested unproved (in comparison to tess4j) approach, it's you call to decide.

File: src/main/java/net/sourceforge/tess4j/util/LoadLibs.java [hunk 10]
```diff
@@ -138,6 +148,25 @@ static void copyResources(URL resourceUrl, File targetPath) throws IOException {
         }
     }
 
+    static void copyFromWarToFolder(VirtualFile virtualFileOrFolder, File targetFolder) throws IOException {
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/util/LoadLibs.java]: @nguyenq great additional method

File: src/main/java/net/sourceforge/tess4j/util/LoadLibs.java [hunk 11]
```diff
@@ -15,24 +15,29 @@
  */
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/util/LoadLibs.java]: @nguyenq now here comes in the changes that are actually in the scope and concern of this merge request.

File: src/main/java/net/sourceforge/tess4j/util/LoadLibs.java [hunk 12]
```diff
@@ -115,8 +121,9 @@ public static synchronized File extractTessResources(String resourceName) {
      * @param resourceUrl
      * @param targetPath
      * @return
+     * @throws URISyntaxException
      */
-    static void copyResources(URL resourceUrl, File targetPath) throws IOException {
+    static void copyResources(URL resourceUrl, File targetPath) throws IOException, URISyntaxException {
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/util/LoadLibs.java]: @nguyenq great additional support for vfs

File: src/main/java/net/sourceforge/tess4j/util/Utils.java [hunk 13]
```diff
@@ -1,79 +1,71 @@
...
-                    }
-                } catch (IllegalAccessException e) {
-                    return String.valueOf(value);
-                }
-            }
-        }
-        return String.valueOf(value);
-    }
-}
+/*
```
Comment (4F2E4A2E) [inline src/main/java/net/sourceforge/tess4j/util/Utils.java]: @nguyenq he is just refactoring some stuff outside the scope of this merge request, with no need for it without reason. This class and logic has been proved it self for many years, i don't see the reason for changing it, it's you call quan.

**Comment by 4F2E4A2E** [comment]
> Fixes problems with resource and memory leaks when loading pdfs with a lot pages. Now loading in memory will be by single page not the whole pdf. (Example: before fix the loading of 200 pages pdf gets more than 3gb ram).
We know that the pdf-to-image support in tess4j is a crappy one, but the API consumer himself should take care of preparing the images in order for tess4j to process it, the scope of converting pdf to images inside tess4j dont't are stop and are not complete by far with the changes you've made, there is a lot more to consider. 

@nguyenq i recommend to merge in the changes of the loadlibs class and that changes only. Everything else are changes being made which can be great or horrible for the api depending on how you see and want to grow and maintain the api, fact is there are no need for the other changes.

**Comment by nguyenq** [comment]
Loading single pages would not cause OOM exceptions anymore but at the expensive cost of setup and teardown of Tesseract engine for every page. The code will need further refactoring to improve the execution speed.

**Pull request merged**